applications/AdjointFluidApplication/custom_utilities/drag_response_function.h

```diff
@@ -0,0 +1,304 @@
+        {
+            // initialize drag flag and element id vectors
+            int k = OpenMPUtils::ThisThread();
+            mElementIds[k] = r_elem.Id() + 1; // force initialization
+            this->GetDragFlagVector(r_elem);
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    void CalculateFirstDerivativesGradient(const Element& rElem,
+                                           const Matrix& rAdjointMatrix,
+                                           Vector& rGradient,
+                                           ProcessInfo& rProcessInfo) override
+    {
+        KRATOS_TRY;
+
+        if (rGradient.size() != rAdjointMatrix.size1())
+            rGradient.resize(rAdjointMatrix.size1(), false);
+
+        Vector& r_drag_flag_vector = this->GetDragFlagVector(rElem);
+        noalias(rGradient) = prod(rAdjointMatrix, r_drag_flag_vector);
+
+        KRATOS_CATCH("");
+    }
+
+    void CalculateSecondDerivativesGradient(const Element& rElem,
+                                            const Matrix& rAdjointMatrix,
+                                            Vector& rGradient,
+                                            ProcessInfo& rProcessInfo) override
+    {
+        KRATOS_TRY;
+
+        if (rGradient.size() != rAdjointMatrix.size1())
+            rGradient.resize(rAdjointMatrix.size1(), false);
+
+        Vector& r_drag_flag_vector = this->GetDragFlagVector(rElem);
+        noalias(rGradient) = prod(rAdjointMatrix, r_drag_flag_vector);
+
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+
+protected:
+    ///@name Protected member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    void CalculateSensitivityGradient(const Element& rElem,
+                                      const Variable<array_1d<double,3>>& rVariable,
+                                      const Matrix& rDerivativesMatrix,
+                                      Vector& rGradient,
+                                      ProcessInfo& rProcessInfo) override
+    {
+        KRATOS_TRY;
+
+        if (rGradient.size() != rDerivativesMatrix.size1())
+            rGradient.resize(rDerivativesMatrix.size1(), false);
+
+        Vector& r_drag_flag_vector = this->GetDragFlagVector(rElem);
+        noalias(rGradient) = prod(rDerivativesMatrix, r_drag_flag_vector);
+
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+
+private:
+    ///@name Member Variables
+    ///@{
+
+    std::string mStructureModelPartName;
+    array_1d<double, TDim> mDragDirection;
+    std::vector<Vector> mDragFlagVector;
+    std::vector<unsigned int> mElementIds;
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    Vector& GetDragFlagVector(const Element& rElement)
+    {
+        int k = OpenMPUtils::ThisThread();
+
+        // if needed, compute the drag flag vector for this element
+        if (rElement.Id() != mElementIds[k])
```

Comment:
i am not clear of what is done here. What are u comparing to what? (the fact that Ids is plural makes me think there is somethign strange ... just discard my comment if iti is fine)